### PR TITLE
support `kw_only` on dataclasses

### PR DESCRIPTION
**Pull request by DetachHead**
## Related issue number

fix #3670 (it's a discussion tho because the issue template didn't let me raise a feature request)
## Checklist

* [x] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [X] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by DetachHead** [comment]
please review

**Comment by DetachHead** [comment]
what's up with the CI on this project? the MR says it's awaiting approval but it's already run? https://github.com/DetachHead/pydantic/actions/runs/1695980747

**Comment by DetachHead** [comment]
how do i make `plugin-fail.txt` expect a different output depending on the python version?

File: pydantic/dataclasses.py [hunk 1]
```diff
@@ -156,8 +158,14 @@ def _process_class(
...
-    cls: Type['Dataclass'] = dataclasses.dataclass(  # type: ignore
-        _cls, init=init, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen
+    cls: Type['Dataclass'] = (
+        dataclasses.dataclass(  # type: ignore
+            _cls, init=init, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen, kw_only=kw_only
+        )
+        if sys.version_info >= (3, 10)
+        else dataclasses.dataclass(  # type: ignore
+            _cls, init=init, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen
+        )
```
Comment (samuelcolvin) [inline pydantic/dataclasses.py:168]: This should be a full `if ...: else:...`, not a "one liner".

File: pydantic/dataclasses.py [hunk 2]
```diff
@@ -199,60 +207,124 @@ def _process_class(
...
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Type[Any] = None,
+            kw_only: bool = False,
+        ) -> Union[Callable[[Type[Any]], Type['Dataclass']], Type['Dataclass']]:
+            ...
+
+else:
+
+    def dataclass(
```
Comment (samuelcolvin) [inline pydantic/dataclasses.py:302]: not sure this should be in `else:`, I think it should be run either with or without `TYPE_CHECKING`?

File: tests/test_dataclasses.py [hunk 3]
```diff
@@ -989,3 +990,17 @@ def __new__(cls, *args, **kwargs):
         instance = cls(a=test_string)
         assert instance._special_property == 1
         assert instance.a == test_string
+
+
+if sys.version_info >= (3, 10):
```
Comment (samuelcolvin) [inline tests/test_dataclasses.py:995]: use `skipIf`, rather than `if` like this.

**Review by samuelcolvin — COMMENTED** [review]
> how do i make `plugin-fail.txt` expect a different output depending on the python version?

One solution (and it would make the overloads much cleaner below), is to always include `kw_only` in the type definition, then just raise a clear error for older versions?

Otherwise LGTM.

**Comment by samuelcolvin** [comment]
please update.

File: pydantic/dataclasses.py (hunk 2 above)
Comment (DetachHead) [reply pydantic/dataclasses.py:302]: unfortunately i can't due to https://github.com/python/mypy/pull/1071. the workaround requires both sides of the `sys.version_info` check to have a "fake" implementation, so if the real implementation wasn't in the `else` mypy would complain that it's a duplicate definition.

i'll add some more comments to make this more clear

Comment (samuelcolvin) [reply pydantic/dataclasses.py:302]: See my message above. Just use the same signature for all versions, then raise an error internally.

Should solve most of these problems.

Comment (DetachHead) [reply pydantic/dataclasses.py:305]: but that way you only get a runtime error. imo it's better to allow type checkers to catch errors like these wherever possible.

if you are happy with this approach, i've made a change to the mypy tests to allow version specific output files which i believe solves that problem

**Comment by duducp** [comment]
Is there any prediction for this feature?

**Comment by KotlinIsland** [comment]
# 🔮
I predict that it will be... BASED!!!

**Comment by samuelcolvin** [comment]
Should go into v1.10 which I'll start working on after we get v1.9.1 out.

File: pydantic/dataclasses.py (hunk 2 above)
Comment (DetachHead) [reply pydantic/dataclasses.py]: https://github.com/python/mypy/pull/10712 has now been merged which solves the issue

**Review by samuelcolvin — APPROVED** [review]

**Comment by samuelcolvin** [comment]
Thanks so much.

**Pull request merged**